File: Dynamic Programming/Java/House Robber II.java
// Problem link: https://leetcode.com/problems/house-robber-ii/description/
// Time: O(n)
// Space: O(1)
class Solution {
    public int rob(int[] nums) {
        if(nums.length == 1)
            return nums[0];
        return Math.max(this.helper(nums, 0, nums.length - 1), this.helper(nums, 1, nums.length));
    }

    private int helper(int[] nums, int left, int right) {
        int one = 0, two = 0, temp;

        for(int i = left; i < right; i++) {
            temp = two;
            two = Math.max(one + nums[i], two);
            one = temp;
        }
        
        return two;
    }
}

File: Dynamic Programming/Python/House Robber II.py
# Problem link: https://leetcode.com/problems/house-robber-ii/description/
# Time: O(n)
# Space: O(1)
class Solution:
    def rob(self, nums: List[int]) -> int:
        
        def helper(left, right):
            one, two = 0, 0

            for i in range(left, right):
                one, two = two, max(one + nums[i], two)
            
            return two
        
        return max(nums[0], helper(0, len(nums) - 1), helper(1, len(nums)))
